fdbclient/NativeAPI.actor.cpp

```diff
@@ -4013,12 +4067,44 @@ ACTOR Future<Void> readVersionBatcher( DatabaseContext *cx, FutureStream<Databas
 	}
 }
 
+std::pair<Version, Standalone<StringRef>> parseVersionStampFromValue(const Value& value) {
+	ASSERT(value.size() >= 10);
+	Version parsedVersion;
+	Standalone<StringRef> parsedVersionstamp = makeString(10);
+	memcpy(&parsedVersion, value.begin(), sizeof(Version));
+	memcpy(mutateString(parsedVersionstamp), value.begin(), 10);
+	return { bigEndian64(parsedVersion), parsedVersionstamp };
+}
+
+ACTOR static Future<Void> getRangeLockSnapshot(DatabaseContext* cx, Version version) {
+	state Future<Void> clientTimeout = delay(5.0);
+	loop {
+		loop choose {
```

Comment:
I suppose this loop is unnecessary.

Reply:
This is intentional so that this loop is retried after line 4092 is triggered. The idea is that if the server returns a snapshot version that is less than the requested version, the client should retry.

Reply:
I am kinda confused. In this case why 

```c++
if (reply.get().version >= version) {
    cx->rangeLockCache.setSnapshot(reply.get().version, reply.get().snapshot);
    return Void();
}
```
won't work?

Reply:
I find out why the proxy returned a version less than the requested one. 8884e39c4 fixes this.

The problem is that txnStateStore recovery applies metadata mutations at version 0. So for range lock cache on proxies, their versions should be "true-up" to the version stored in the rangeLockVersionKey.